src/details/ArborX_DetailsDistributor.hpp

```diff
@@ -62,6 +62,91 @@ struct Result
+                      std::vector<int> &unique_ranks, std::vector<int> &counts,
+                      std::vector<int> &offsets)
+{
+  ARBORX_ASSERT(unique_ranks.empty());
+  ARBORX_ASSERT(offsets.empty());
+  ARBORX_ASSERT(counts.empty());
+  ARBORX_ASSERT(permutation_indices.extent_int(0) ==
+                lastElement(batched_offsets));
+  ARBORX_ASSERT(batched_ranks.size() + 1 == batched_offsets.size());
+  static_assert(
+      std::is_same<typename InputView::non_const_value_type, int>::value, "");
+  static_assert(std::is_same<typename OutputView::value_type, int>::value, "");
+
+  offsets.push_back(0);
+
+  auto const n = batched_ranks.size();
+  if (n == 0 || lastElement(batched_offsets) == 0)
+    return;
+
+  using DeviceType = typename InputView::traits::device_type;
+  using ExecutionSpace = typename InputView::traits::execution_space;
+  auto offset_ranks_copy = clone(batched_offsets);
+  auto unique_ranks_copy = clone(batched_ranks);
+  Kokkos::View<int, DeviceType, Kokkos::MemoryTraits<Kokkos::Atomic>> position(
+      "position");
+  Kokkos::deep_copy(position, 1);
+
+  Kokkos::parallel_for(
+      ARBORX_MARK_REGION("find_unique_starts"),
+      Kokkos::RangePolicy<ExecutionSpace>(0, batched_ranks.size() + 1),
+      KOKKOS_LAMBDA(unsigned int i) {
+        if (i == batched_ranks.size())
+        {
+          auto const my_position = position()++;
+          offset_ranks_copy(my_position) = batched_offsets(i);
+        }
+        else if (i > 0 && batched_ranks(i) != batched_ranks(i - 1))
+        {
+          auto const my_position = position()++;
+          unique_ranks_copy(my_position) = batched_ranks(i);
+          offset_ranks_copy(my_position) = batched_offsets(i);
+        }
+      });
+
+  auto const host_position =
+      Kokkos::create_mirror_view_and_copy(Kokkos::HostSpace(), position);
+  auto sorted_offsets = InputView(
+      offset_ranks_copy, std::make_pair(0, static_cast<int>(host_position())));
+  auto const permutation = ArborX::Details::sortObjects(sorted_offsets);
+  auto sorted_unique_ranks = InputView(
+      unique_ranks_copy, std::make_pair(0, static_cast<int>(host_position())));
+  ArborX::Details::applyPermutation(permutation, sorted_unique_ranks);
+
+  auto const unique_ranks_host = Kokkos::create_mirror_view_and_copy(
+      Kokkos::HostSpace(), sorted_unique_ranks);
+  unique_ranks.reserve(unique_ranks_host.size() - 1);
+
+  auto const offsets_host =
+      Kokkos::create_mirror_view_and_copy(Kokkos::HostSpace(), sorted_offsets);
+  offsets.reserve(sorted_offsets.size());
+  counts.reserve(sorted_offsets.size() - 1);
+  for (unsigned int i = 1; i < sorted_offsets.size(); ++i)
+  {
+    int const count = offsets_host(i) - offsets_host(i - 1);
+    if (count > 0)
+    {
+      offsets.push_back(offsets_host(i));
+      counts.push_back(count);
+      unique_ranks.push_back(unique_ranks_host(i - 1));
+    }
+  }
+
+  ArborX::iota(permutation_indices, 0);
```

Comment:
Why do that? This should not be necessary. We should not use permutation_indices at all, both inside and outside of this function.